### ISSUE TITLE
Writing to an element of an array-typed field requires a read privilege

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#436]
The compiler complains that task `bar` in the following snippet requires a read privilege on `fld`.

```
import 'regent'

struct Foo {
  fld : int[3];
}

task bar(r : region(ispace(int1d),Foo))
where
  writes(r.fld)
do
  for elem in r do
    elem.fld[0] = 12
  end
end
```

**Comment by elliottslaughter** [comment]
Without field-slicing the array, this is fundamentally the only way we can do this. There is no way to write to only part of a field in Legion.

@magnatelee and I have talked about field slicing arrays in the past, but I've always shied away from it because it's not possible in general to compute a single base and stride for all the fields. (You can do it with a sufficiently restrictive set of layout constraints, but that's not something we've been comfortable with doing so far.)

So, unfortunately a wont-fix unless we decide to go after field slicing of arrays.

**Comment by lightsighter** [comment]
@elliottslaughter What are the implications on this for code generation? Does this mean that you're actually reading out the whole array, writing the value, and then writing the whole array back? Or do you guys do something more clever like get a pointer to the array in memory and then only write the one value in the proper location?

**Comment by elliottslaughter** [comment]
@magnatelee will have to confirm, but glancing through the code it looks like `elem.fld` returns a reference to the entire array, and then if you assign to `elem.fld[0]` then you're basically doing a standard Terra array access to a reference. I don't think the code generator knows about arrays at all.

**Comment by elliottslaughter** [comment]
I'm not sure what @lightsighter is suggesting here. The error @manopapad receives above is correct: `reads` is required but wasn't specified. The issue of inference seems orthogonal to this.

**Comment by lightsighter** [comment]
@elliottslaughter Why is a read privilege necessary here? The code only writes to the field (to be precise it's only writing to a part of the field) but it definitely isn't reading the field at all. I fail to see why a read privilege is necessary. I'm assuming this is a bug in your privilege inference analysis.

**Comment by elliottslaughter** [comment]
@magnatelee With field sliced arrays is this now fixed?

**Comment by magnatelee** [comment]
@elliottslaughter @manopapad can rewrite the field space to use field sliced arrays if he wants to, but he would still need to set the right field constraint so that those sliced elements are colocated. (Otherwise, performance will be horrible.) At the moment, fields constraints must be specified on per-task basis, which might be painful to do in big programs like Soleil.

**Comment by magnatelee** [comment]
@elliottslaughter with the fix for #586 write accesses to array-type fields no longer require a read privilege, and we know write-only privileges are promoted to read-write privileges. So, can we close this issue?

**Comment by manopapad** [comment]
Yup
